src/puppetlabs/puppetdb/mq_listener.clj

```diff
@@ -167,154 +158,88 @@
                   :trace     (map str (.getStackTrace e))}]
     (update-in msg [:annotations :attempts] conj attempt)))
 
-(defn wrap-with-exception-handling
-  "Wrap a message processor `f` such that all Throwable or `fatal?`
-  exceptions are caught.
-
-  If a `fatal?` exception is thrown, the supplied `on-fatal` function
-  is invoked with the message and the exception as arguments.
+;; The number of times a message can be retried before we discard it
+(def maximum-allowable-retries 16)
 
-  If any other Throwable exception is caught, the supplied `on-retry`
-  function is invoked with the message and the exception as
-  arguments."
-  [f on-retry on-fatal]
+(defn mark-both-metrics!
+  "Calls `mark!` on the global and command specific metric for `k`"
+  [command version k]
+  (mark! (global-metric k))
+  (mark! (cmd-metric command version k)))
+
+(defn update-both-metrics!
+  "Calls `update!` on the global and command specific metric for `k`"
+  [command version k v]
+  (update! (global-metric k) v)
+  (update! (cmd-metric command version k) v))
+
+(defn call-with-command-metrics
+  "Invokes `f` including the related metrics updates"
+  [command version retries f]
+  (create-metrics-for-command! command version)
+
+  (mark-both-metrics! command version :seen)
+  (update-both-metrics! command version :retry-counts retries)
+
+  (mutils/multitime!
+   [(global-metric :processing-time)
+    (cmd-metric command version :processing-time)]
+   (f))
+
+  (mark-both-metrics! command version :processed))
+
+(defn message-handler-with-retries
+  "This function processing of the message, retrying messages that
+  fail and discarding messages that have fatal errors or have exceeded
+  their maximum allowed attempts. `delay-message-fn` and
+  `discard-message-fn` are both functions of two arguments, a
+  `message` and an `exception`. `process-message-fn` is a function
+  that accepts a message as it's argument"
+  [delay-message-fn discard-message-fn process-message-fn]
   (fn [msg]
     (try+
 
-     (mark! (global-metric :seen))
-     (time! (global-metric :processing-time)
-            (f msg))
-     (mark! (global-metric :processed))
-
-     (catch fatal? {:keys [cause]}
-       (on-fatal msg cause)
-       (mark! (global-metric :fatal)))
+     (let [{:keys [command version annotations] :as parsed-result} (parse-command msg)
```

Comment:
As we discussed, we might change the arguments to the callbacks, or we could wait until the next revision.